drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,319 @@
+				Description:              to.StringPtr("User requested port to be accessible from Internet via docker-machine"),
+				SourceAddressPrefix:      to.StringPtr("*"),
+				DestinationAddressPrefix: to.StringPtr("*"),
+				SourcePortRange:          to.StringPtr("*"),
+				DestinationPortRange:     to.StringPtr(openPort.Port),
+				Access:                   network.SecurityRuleAccessAllow,
+				Direction:                network.SecurityRuleDirectionInbound,
+				Protocol:                 openPort.Proto,
+			},
+		}
+	}
+	// Log open ports to be deleted and added
+	if len(openPortsToBeDeleted) > 0 || len(openPortsToBeAdded) > 0 {
+		removedPorts := make([]string, len(openPortsToBeDeleted))
+		addedPorts := make([]string, len(openPortsToBeAdded))
+		i := 0
+		for openPort := range openPortsToBeDeleted {
+			removedPorts[i] = openPort
+		}
+		i = 0
+		for openPort := range openPortsToBeAdded {
+			addedPorts[i] = openPort
+		}
+		log.Info("Updating security rule configuration", logutil.Fields{
+			"openPortsRemoved": removedPorts,
+			"openPortsAdded":   addedPorts,
+		})
+	}
+	// Remove all unnecessary open ports and release the priorities
+	for _, rule := range openPortsToBeDeleted {
+		delete(takenPriorities, to.Int32(rule.Priority))
+		rules = removeSecurityRule(rules, rule)
+	}
+	// Add any necessary open ports
+	newRules := *rules
+	priority := int32(1000)
+	for _, rule := range openPortsToBeAdded {
+		// Find a priority that can be given to this rule
+		_, taken := takenPriorities[priority]
+		for taken {
+			priority++
+			_, taken = takenPriorities[priority]
+		}
+		rule.Priority = to.Int32Ptr(priority)
+		priority++
+		newRules = append(newRules, rule)
+	}
+	return &newRules, nil
+}
+
+// checkIfDockerConflict does a rudimentary check to see if the rule could potentially conflict with the ability
+// for the docker port to be accessible. If the rule does not pass all rudimentary checks, it returns an error.
+func checkIfDockerConflict(rule network.SecurityRule, dockerPort int) error {
+	f := logutil.Fields{
+		"name":        to.String(rule.Name),
+		"description": to.String(rule.Description),
+		"priority":    to.Int32(rule.Priority),
+		"access":      rule.Access,
+		"direction":   rule.Direction,
+		"protocol":    rule.Protocol,
+		"portRanges":  getDestinationPorts(rule),
+	}
+	priority := to.Int32(rule.Priority)
+	if priority > dockerPriority {
+		return nil
+	}
+	purpose := strings.ToLower(fmt.Sprintf("%s %s %s", rule.Access, rule.Direction, rule.Protocol))
+	if purpose == "deny inbound tcp" || purpose == "deny inbound *" {
+		// check if docker port is covered by rule
+		for _, pr := range getDestinationPorts(rule) {
+			portWithinRange, err := checkIfPortInPortRange(dockerPort, pr)
+			if err != nil {
+				return fmt.Errorf("%s: %s", err, f)
+			}
+			if portWithinRange {
+				return fmt.Errorf("rule %s set to %s requests to the Docker port %d: %s",
+					to.String(rule.Name), purpose, dockerPort, f)
+			}
+		}
+	}
+	return nil
+}
```

Comment:
We currently fail on create if a user has any existing security rule that has all the following:
- a higher priority than the DockerAllowAny rule (i.e. between 101 <= priority <= 299)
- Denies inbound tcp/* requests
- Contains the dockerPort in the DestinationPortRanges

Examples (log statements may be outdated):
![blocking_docker_port_with_rules_0](https://user-images.githubusercontent.com/13127443/77803230-e2e00380-7039-11ea-8c27-d3541a7dea47.png)
![blocking_docker_port_with_rules_1](https://user-images.githubusercontent.com/13127443/77803236-e4113080-7039-11ea-8fe7-3be99aab8b21.png)
![blocking_docker_port_with_rules_2](https://user-images.githubusercontent.com/13127443/77803240-e5425d80-7039-11ea-9390-373b0326653d.png)
![blocking_docker_port_with_rules_3](https://user-images.githubusercontent.com/13127443/77803244-e5daf400-7039-11ea-9dc1-52786206bc31.png)

Such rules would would conflict with the way we set the Docker rule settings, shown below:

```
# Output from the r.DockerRule.Rule(), where r is the RancherMachineSecurityRuleConfiguration

network.SecurityRule{
 	Name: to.StringPtr("DockerAllowAny"),
 	SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
 		Description:              to.StringPtr("Allow docker engine access (TLS-protected)"),
 		SourceAddressPrefix:      to.StringPtr("*"),
 		DestinationAddressPrefix: to.StringPtr("*"),
 		SourcePortRange:          to.StringPtr("*"),
 		DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", dockerPort)),
 		Access:                   network.SecurityRuleAccessAllow,
 		Direction:                network.SecurityRuleDirectionInbound,
 		Protocol:                 network.SecurityRuleProtocolTCP,
 		Priority:                 to.Int32Ptr(300),
 	},
 }
```

However, based on the [`SecurityRulePropertiesFormat` Azure docs](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-11-01/network?tab=doc#SecurityRulePropertiesFormat), the user could theoretically have a rule that denies inbound tcp requests to the dockerPort (and would fail on create like the above situations), but only to specific:
- SourcePortRange(s)
- SourceAddressPrefix(es) / DestinationAddressPrefix(es)
- SourceApplicationSecurityGroups(s) / DestinationAddressSecurityGroup(s)

Would it make sense to accommodate for these use cases by logging a warning instead of returning an error? This would result on a post-provision failure while trying to set up Docker instead of a on provision failure.

Reply:
No longer relevant; based on https://github.com/rancher/machine/pull/77#issuecomment-625564555 the user is expected to provide a NSG with the required security rules, so this check is omitted.